File: 001050/22.java
/**
2ms。。。有点刺激
 */
class Solution {
    
    private int n;
    private List<String> ans;
    
    private void backtrack(String str, int left, int right) {
        if (str.length() == 2 * this.n) {
            this.ans.add(str);
            return ;
        }
        if (left < this.n) {
            backtrack(str+"(", left+1, right);
        }
        if (left > right) {
            backtrack(str+")", left, right+1);
        }
    }
    
    public List<String> generateParenthesis(int n) {
        this.ans = new ArrayList<String>();
        this.n = n;
        
        backtrack("", 0, 0);
        
        return this.ans;
    }
    
}

File: regular_season/22.py
'''
给出 n 代表生成括号的对数，请你写出一个函数，使其能够生成所有可能的并且有效的括号组合。
'''
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        
        def bt(s = '', left = 0, right = 0):
            if len(s) == 2 * n:
                ans.append(s)
                return
            if left < n:
                bt(s+'(', left+1, right)
            if left > right:
                bt(s+')', left, right+1)
        
        bt()
        return ans